src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -2785,5 +2837,1166 @@ private void AssertValid()
+            {
+                return int.TrailingZeroCount(value._sign);
+            }
+
+            ulong result = 0;
+
+            if (value._sign >= 0)
+            {
+                // When the value is positive, we simply need to do a tzcnt for all bits until we find one set
+
+                uint part = value._bits[0];
+
+                for (int i = 1; (part == 0) && (i < value._bits.Length); i++)
+                {
+                    part = value._bits[i];
+                    result += (sizeof(uint) * 8);
+
+                    i++;
+                }
+
+                result += uint.TrailingZeroCount(part);
+            }
+            else
+            {
+                // When the value is negative, we need to tzcnt the two's complement representation
+                // We'll do this "inline" to avoid needing to unnecessarily allocate.
+
+                uint part = ~value._bits[0] + 1;
+
+                for (int i = 1; (part == 0) && (i < value._bits.Length); i++)
+                {
+                    // Simply process bits, adding the carry while the previous value is zero
+
+                    part = ~value._bits[i] + 1;
+                    result += (sizeof(uint) * 8);
+
+                    i++;
+                }
+
+                result += uint.TrailingZeroCount(part);
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.GetShortestBitLength()" />
+        long IBinaryInteger<BigInteger>.GetShortestBitLength()
+        {
+            uint[]? bits = _bits;
+
+            if (bits is null)
+            {
+                int value = _sign;
+
+                if (value >= 0)
+                {
+                    return (sizeof(int) * 8) - int.LeadingZeroCount(value);
+                }
+                else
+                {
+                    return (sizeof(int) * 8) + 1 - int.LeadingZeroCount(~value);
+                }
+            }
+
+            long result = (bits.Length - 1) * 32;
+
+            if (_sign >= 0)
+            {
+                result += (sizeof(uint) * 8) - uint.LeadingZeroCount(bits[^1]);
+            }
+            else
+            {
+                result += (sizeof(uint) * 8) + 1 - uint.LeadingZeroCount(~bits[^1]);
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.GetByteCount()" />
+        int IBinaryInteger<BigInteger>.GetByteCount()
+        {
+            uint[]? bits = _bits;
+
+            if (bits is null)
+            {
+                return sizeof(int);
+            }
+            return bits.Length * 4;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.TryWriteLittleEndian(Span{byte}, out int)" />
+        bool IBinaryInteger<BigInteger>.TryWriteLittleEndian(Span<byte> destination, out int bytesWritten)
+        {
+            uint[]? bits = _bits;
+            int byteCount = (bits is null) ? sizeof(int) : bits.Length * 4;
```

Comment:
Possible integer overflow.

Reply:
I'm going to add enforcement that `bits.Length` isn't too long, otherwise this code gets massively more complicated.

Reply:
Fixed by adding validation that `bits.Length <= (Array.MaxLength / sizeof(uint));`